ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java

```diff
@@ -34,52 +36,87 @@
 public class ForkIdManager {
 
   private final Hash genesisHash;
-  private final List<ForkId> forkIds;
+  private final List<ForkId> blockNumbersForkIds;
+  private final List<ForkId> timestampsForkIds;
 
-  private final List<Long> forkBlockNumbers;
-  private final LongSupplier chainHeadSupplier;
+  private final List<Long> blockNumberForks;
+  private final List<Long> timestampForks;
+
+  private final Supplier<BlockHeader> chainHeadSupplier;
   private final long forkNext;
   private final boolean onlyZerosForkBlocks;
   private final long highestKnownFork;
   private Bytes genesisHashCrc;
   private final boolean legacyEth64;
 
   public ForkIdManager(
-      final Blockchain blockchain, final List<Long> nonFilteredForks, final boolean legacyEth64) {
+      final Blockchain blockchain,
+      final List<Long> blockNumberForks,
+      final List<Long> timestampForks,
+      final boolean legacyEth64) {
     checkNotNull(blockchain);
-    checkNotNull(nonFilteredForks);
-    this.chainHeadSupplier = blockchain::getChainHeadBlockNumber;
+    checkNotNull(blockNumberForks);
+    this.chainHeadSupplier = blockchain::getChainHeadHeader;
     this.genesisHash = blockchain.getGenesisBlock().getHash();
-    this.forkIds = new ArrayList<>();
+    this.blockNumbersForkIds = new ArrayList<>();
+    this.timestampsForkIds = new ArrayList<>();
     this.legacyEth64 = legacyEth64;
-    this.forkBlockNumbers =
-        nonFilteredForks.stream()
+    this.blockNumberForks =
+        blockNumberForks.stream()
             .filter(fork -> fork > 0L)
             .distinct()
             .sorted()
             .collect(Collectors.toUnmodifiableList());
-    this.onlyZerosForkBlocks = nonFilteredForks.stream().allMatch(value -> 0L == value);
+    this.timestampForks =
+        timestampForks.stream()
+            .filter(fork -> fork > 0L)
```

Comment:
should this be >= 0L ?  for a merge-at-genesis ?  https://github.com/hyperledger/besu/pull/4792 is allowing for merge on genesis block, with no transition schedule (which presumably would be extended to include a timestamp protocol schedule at time 0 also)

Reply:
That's one for https://github.com/hyperledger/besu/pull/4815/commits/8b5e3ed1092bdd2b58851ba123976268a41044bf cc @jframe